src/nupic/types/SdrProxy.hpp

```diff
@@ -143,17 +161,248 @@ class SDR_Proxy : public SDR
-        { NTA_THROW << _SDR_Proxy_setter_error_message; }
-    void setSDR( const SparseDistributedRepresentation &value ) override
-        { NTA_THROW << _SDR_Proxy_setter_error_message; }
+/**
+ * TODO DOCUEMNTATION
+ */
+class SDR_Concatenation : public SDR_ReadOnly_
+{
+protected:
+    UInt         axis_;
+    vector<SDR*> inputs_;
+    vector<UInt> callback_handles_;
+    vector<UInt> destroyCallback_handles_;
+    mutable bool dense_valid_lazy;
+
+    void clear() const override {
+        SDR::clear();
+        // Always advertise that this SDR has dense data.
+        dense_valid = true;
+        // But make note that this SDR does not actually have dense data, it
+        // will be computed it when it's requested.
+        dense_valid_lazy = false;
+    }
+
+    void deconstruct() override {
+        // Unlink everything at death.
+        for(auto i = 0u; i < inputs_.size(); i++) {
+            inputs_[i]->removeCallback( callback_handles_[i] );
+            inputs_[i]->removeDestroyCallback( destroyCallback_handles_[i] );
+        }
+        // Clear internal data.
+        inputs_.clear();
+        callback_handles_.clear();
+        destroyCallback_handles_.clear();
+        dense_valid_lazy = false;
+        // Notify SDR parent class.
+        SDR::deconstruct();
+    }
+
+public:
+    const UInt         &axis   = axis_;
+    const vector<SDR*> &inputs = inputs_;
+
+    SDR_Concatenation(SDR &inp1, SDR &inp2, UInt axis=0u)
+        { initialize({    &inp1,     &inp2},     axis); }
+    SDR_Concatenation(SDR &inp1, SDR &inp2, SDR &inp3, UInt axis=0u)
+        { initialize({    &inp1,     &inp2,     &inp3},     axis); }
+    SDR_Concatenation(SDR &inp1, SDR &inp2, SDR &inp3, SDR &inp4, UInt axis=0u)
+        { initialize({    &inp1,     &inp2,     &inp3,     &inp4},     axis); }
+
+    SDR_Concatenation(vector<SDR*> inputs, UInt axis=0u)
+        { initialize(inputs, axis); }
+
+    void initialize(const vector<SDR*> inputs, const UInt axis=0u)
+    {
+        NTA_CHECK( inputs.size() >= 1u )
+            << "Not enough inputs to SDR_Concatenation, need at least 2 SDRs got " << inputs.size() << ".";
+        inputs_.assign( inputs.begin(), inputs.end() );
+        axis_ = axis;
+        const UInt n_dim = inputs[0]->dimensions.size();
+        NTA_CHECK( axis_ < n_dim );
+        // Determine dimensions & check input dimensions.
+        vector<UInt> dims = inputs[0]->dimensions;
+        dims[axis] = 0;
+        for(auto i = 0u; i < inputs.size(); ++i) {
+            NTA_CHECK( inputs[i]->dimensions.size() == n_dim )
+                << "All inputs to SDR_Concatenation must have the same number of dimensions!";
+            for(auto d = 0u; d < n_dim; d++) {
+                if( d == axis )
+                    dims[axis] += inputs[i]->dimensions[d];
+                else
+                    NTA_CHECK( inputs[i]->dimensions[d] == dims[d] )
+                        << "All dimensions except the axis must be the same! "
```

Comment:
> All dimensions except the axis must be the same!

This is the restriction that I have the most trouble with.  In the case of NetworkAPI links, a user could connect multiple links from any two or more regions to a single region.  The buffers need to be concatenated.  But the buffers may not be the same number of dimensions (although I can add 1's), the same number of elements, or the same dimensions on any axis.  The data could actually be in different data types although I will convert each to the type of the destination buffer first.

I have this working by flatting the buffers and copying each adjacent to each other. The resulting buffer's width is the sum of the number of elements in all incoming buffers.  Of course the dimensions become 1D which is the problem.

I guess the point is that I would not be able to use this for NetworkAPI with these restrictions.  It probably would be useful for other things.

Reply:
@dkeeney,
Thank you for reviewing this PR.

> This is the restriction that I have the most trouble with.

This restriction exists so that the resulting SDR has a rectangular shape.  The quick fixes are to either flatten the SDRs to 1D or add dimension of size 1.

I have an idea about how to properly fix this restriction and -no surprises- it involves SDR subclasses.  First we will need a subclass which reshapes SDR's and takes topology into account.  The current SDR_Proxy/reshape is insufficient because it simply changes the dimensions.  This new reshaping SDR will need to actually move around all of the bits.  This movement amounts to some index math so it should be pretty fast & space efficient.

Example Pseudo Code:
```Python
A = SDR( 100 x 100 x 3 )
B = SDR( 10 x 10 x 44 )
# Can not concatenate A & B, because no dimensions match.
# The next line will divide SDR A into blocks of [10 x 10 x 3]
# and then copy each block to output C in shape [1 x 1 x 300]
C = SDR_Reshape( A, 10 x 10 x 300 )
# Now we can concatenate B & C, because the dimensions match.
D = Concatenate( B, C, axis=2 )
D.dimensions == [10, 10, 344]
```

Reply:
I like this 👍

I have fixed the Link so it now concatenates and retains dimensions as long as  the lower dimensions match (one axis).  This works for non-SDR Arrays as well as SDRs.  This is all done before any Array buffers (or SDRs) are created.